### ISSUE TITLE
STJ source generator fails when defining the serializer context in 2 partial classes

### ISSUE THREAD

**Issue by hugoqribeiro** [dotnet/runtime#99669]
### Description

If you define a serializer context (`JsonSerializerContext`) in 2 partial classes, the source generator fails with errors and a warning like this:

```
Warning	CS8785	Generator 'JsonSourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentException' with message 'The hintName 'SerializerContext.MyClass1.g.cs' of the added source file must be unique within a generator.
```

We are doing this to support the following scenario:

- The first partial class is produced by a code generator based on metadata.
- We let developers add the second partial class to include classes in the serialized that are not know to the code generator.

### Reproduction Steps

Define a set of classes that need to be serialized:

```
public class MyClass1
{
    // (...)
}

public class MyClass2
{
    // (...)
}
```

Define a serializer context in a partial class:

```
[JsonSerializable(typeof(MyClass1))]
internal partial class SerializerContext : JsonSerializerContext
{
}
```

Define another partial class for the same serializer:

```
[JsonSerializable(typeof(MyClass2))]
internal partial class SerializerContext
{
}
```

> This is easily reproducible in a .NET 8 console app.

### Expected behavior

The source generator should be able to generate without errors:

### Actual behavior

The following errors are produced when building:

```
warning CS8785: Generator 'JsonSourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentException' with message 'The hintName 'SerializerContext.MyClass1.g.cs' of the added source file must be unique within a generator.

error CS0534: 'SerializerContext' does not implement inherited abstract member 'JsonSerializerContext.GetTypeInfo(Type)'

error CS0534: 'SerializerContext' does not implement inherited abstract member 'JsonSerializerContext.GeneratedSerializerOptions.get'
```

### Regression?

This worked correctly in .NET 7.

### Known Workarounds

Can't seam to find one...

### Configuration

.NET 8.0.102

**Comment by martincostello** [comment]
Does it work if you only put the attribute on one of the two partial declarations?

**Comment by eiriktsarpalis** [comment]
I would inclined to ascribe this to by-design behavior. The source generator is triggered using `JsonSerializableAttribute` annotations, so if the same class declares attributes in multiple locations then generator would be triggered multiple times for the same context type. It wouldn't be easy to reconcile this without compromising the incremental caching properties of the generator.

**Comment by hugoqribeiro** [comment]
Have you considered that this worked in version 7.x?

And this is done with code generation, which implies having 2 locations (the one generated, and the one added by the developer)?

**Comment by eiriktsarpalis** [comment]
You're correct that this used to work in earlier versions. The current behaviour is a direct result of https://github.com/dotnet/runtime/pull/86616, which compiles separate `JsonSerializerContext` declarations independently. This is motivated out of IDE performance concerns: the source generator should only emit code for the context you are currently editing. The drawback of that approach is that it's not possible to group incremental values by a given type without forcing re-evaluation of the entire graph, so the source generator will generate code twice for the same type if `JsonSerializable` are present in multiple declarations.

Given that this is a tooling-side regression impacting a rare use case that has a clear workaround, I don't think we should attempt to service this. At the same time, I'll reopen the issue so that we can try improving the situation in the future. For this though, I believe that we would require grouping functionality for the `IncrementalValuesProvider<T>` type in Roslyn.

**Comment by aries-zhang** [comment]
In my use case there are mainly 2 groups of types that need to be serialized/deserialized. Group 1 is those built-in types, such as `Dictionary<string, string>`, etc. Group 2 is those classes declared in my assembly, such as Class1, Class2, etc. 

It is cumbersome to having to declare `[JsonSerializable(typeof(Class1))]` for every type that needs to be serde'd in a centralized place. Easier to mark these classes in place where they are declared, so I can have a source generator to generate those attributes, so it ends up in:

```
[HeyThisNeedsToBeSerialized]
public class Class1 {
    ...
}

[JsonSerializable(typeof(Dictionary<string, string>))] // <-- Types that can't be modified
internal partial class SerializerContext : JsonSerializerContext
{
}

[JsonSerializable(typeof(Class1))] // <-- types that can be declared as serializable in-place
[JsonSerializable(typeof(Class2))] 
partial class SerializerContext
{
}
```
but this issue makes it impossible to do this. 

Although wondering if I'm stupidly missing something and there is an easier way to achieve the same?
